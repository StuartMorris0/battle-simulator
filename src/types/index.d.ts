declare interface PlayerHealth {
  current: number;
  lost: number | undefined;
}

declare interface Player {
  playerName: string;
  dice1: number;
  dice2: number;
  health: PlayerHealth;
  hasWon: boolean;
}
