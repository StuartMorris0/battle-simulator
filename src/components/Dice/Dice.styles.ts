import { motion } from 'framer-motion';
import styled from 'styled-components';

export const RollingDice = styled(motion.div)`
  margin: 4px;
  width: 50px;
  height: 50px;
  background: black;
  border-radius: 10px;
  overflow: hidden;
  & img {
    width: 50px;
    height: 50px;
  }
`;

export const Rolled = styled(motion.div)`
  display: block;

  & img {
    width: 50px;
    height: 50px;
  }
`;
