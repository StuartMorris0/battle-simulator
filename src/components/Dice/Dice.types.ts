export interface IDiceProps {
  value: number | undefined;
  isRolling: boolean;
}
