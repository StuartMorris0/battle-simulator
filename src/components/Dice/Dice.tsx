import React from 'react';

import blank from 'assets/icons/dice/blank.svg';
import five from 'assets/icons/dice/five.svg';
import four from 'assets/icons/dice/four.svg';
import one from 'assets/icons/dice/one.svg';
import six from 'assets/icons/dice/six.svg';
import three from 'assets/icons/dice/three.svg';
import two from 'assets/icons/dice/two.svg';

import { Rolled, RollingDice } from './Dice.styles';
import { IDiceProps } from './Dice.types';

const getImageSource = (value: number) => {
  switch (value) {
    case 1:
      return one;
    case 2:
      return two;
    case 3:
      return three;
    case 4:
      return four;
    case 5:
      return five;
    case 6:
      return six;
    default:
      return blank;
  }
};

const Dice: React.FC<IDiceProps> = ({ value, isRolling }: IDiceProps) => {
  if (isRolling) {
    return (
      <RollingDice
        data-testid="RollingDice"
        animate={{ scale: [1, 0.75, 0.5, 0.5, 1], rotate: [0, 0, 270, 270, 0], borderRadius: ['20%', '20%', '50%', '50%', '20%'] }}
        transition={{ repeat: Infinity, ease: 'linear', duration: 1.5 }}
      >
        <img src={blank} alt="Rolling..." />
      </RollingDice>
    );
  }

  if (value !== undefined) {
    const imageSrc = getImageSource(value);
    return (
      <Rolled data-testid="RolledDice" animate={{}} whileHover={{ scale: 1.2 }}>
        <img src={imageSrc} alt="dice" />
      </Rolled>
    );
  }
  return null;
};

export default Dice;
