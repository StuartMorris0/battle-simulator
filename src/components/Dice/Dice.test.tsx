import React from 'react';

import { render } from 'global/utils/test-utils';

import Dice from './Dice';
import { IDiceProps } from './Dice.types';

describe('Dice Component', () => {
  let props: IDiceProps;
  beforeEach(() => {
    props = {
      value: 2,
      isRolling: false,
    };
  });

  const renderComponent = () => render(<Dice {...props} />);

  test('renders rolled dice', () => {
    const { getByTestId, getByRole } = renderComponent();
    expect(getByTestId('RolledDice')).toBeInTheDocument();
    expect(getByRole('img')).toBeInTheDocument();
  });

  test('renders rolling blank dice', () => {
    props.isRolling = true;
    const { getByTestId, getByRole } = renderComponent();
    expect(getByTestId('RollingDice')).toBeInTheDocument();
    expect(getByRole('img')).toBeInTheDocument();
  });

  test('renders empty when in starting state', () => {
    props.value = undefined;
    const { container } = renderComponent();
    expect(container).toBeEmptyDOMElement();
  });

  it.each([1, 2, 3, 4, 5, 6, 7])('snapshot test for variant %s', (variant) => {
    // 7 will render the blank dice image
    props.value = variant;
    const { container } = renderComponent();
    expect(container.firstChild).toMatchSnapshot();
  });
});
