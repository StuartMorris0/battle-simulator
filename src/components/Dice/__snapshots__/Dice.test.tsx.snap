// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dice Component snapshot test for variant 1 1`] = `
.c0 {
  display: block;
}

.c0 img {
  width: 50px;
  height: 50px;
}

<div
  class="c0"
  data-testid="RolledDice"
>
  <img
    alt="dice"
    src="one.svg"
  />
</div>
`;

exports[`Dice Component snapshot test for variant 2 1`] = `
.c0 {
  display: block;
}

.c0 img {
  width: 50px;
  height: 50px;
}

<div
  class="c0"
  data-testid="RolledDice"
>
  <img
    alt="dice"
    src="two.svg"
  />
</div>
`;

exports[`Dice Component snapshot test for variant 3 1`] = `
.c0 {
  display: block;
}

.c0 img {
  width: 50px;
  height: 50px;
}

<div
  class="c0"
  data-testid="RolledDice"
>
  <img
    alt="dice"
    src="three.svg"
  />
</div>
`;

exports[`Dice Component snapshot test for variant 4 1`] = `
.c0 {
  display: block;
}

.c0 img {
  width: 50px;
  height: 50px;
}

<div
  class="c0"
  data-testid="RolledDice"
>
  <img
    alt="dice"
    src="four.svg"
  />
</div>
`;

exports[`Dice Component snapshot test for variant 5 1`] = `
.c0 {
  display: block;
}

.c0 img {
  width: 50px;
  height: 50px;
}

<div
  class="c0"
  data-testid="RolledDice"
>
  <img
    alt="dice"
    src="five.svg"
  />
</div>
`;

exports[`Dice Component snapshot test for variant 6 1`] = `
.c0 {
  display: block;
}

.c0 img {
  width: 50px;
  height: 50px;
}

<div
  class="c0"
  data-testid="RolledDice"
>
  <img
    alt="dice"
    src="six.svg"
  />
</div>
`;

exports[`Dice Component snapshot test for variant 7 1`] = `
.c0 {
  display: block;
}

.c0 img {
  width: 50px;
  height: 50px;
}

<div
  class="c0"
  data-testid="RolledDice"
>
  <img
    alt="dice"
    src="blank.svg"
  />
</div>
`;
